File: backend/routes/modrinth.py
"""Modrinth API routes."""
import os
from pathlib import Path
from flask import Blueprint, jsonify, request

from backend.services import server_storage
from backend.services.modrinth_client import ModrinthClient
from backend.services.server_registry import get_server_process_registry

modrinth_bp = Blueprint('modrinth', __name__, url_prefix='/api/modrinth')
modrinth_client = ModrinthClient()
process_registry = get_server_process_registry()


def _resolve_mods_folder(server_id: str | None):
    if server_id:
        server = server_storage.get_server(server_id)
        if not server:
            return None, ({'error': 'Server not found'}, 404)
        try:
            path = process_registry.resolve_mods_path(server)
            return path, None
        except ValueError as exc:
            return None, ({'error': str(exc)}, 400)

    legacy_root = Path(os.path.join(os.getcwd(), 'server'))
    mods_path = legacy_root / 'mods'
    mods_path.mkdir(parents=True, exist_ok=True)
    return mods_path, None


@modrinth_bp.route('/search', methods=['GET'])
def search_mods():
    """
    Search for mods on Modrinth.
    
    Query params:
        - query: Search text
        - mc_version: Minecraft version (e.g., "1.20.1")
        - loader: Mod loader (e.g., "fabric", "forge")
        - limit: Results per page (default 20, max 100)
        - offset: Pagination offset
        - index: Sort by (downloads, relevance, follows, newest, updated)
    """
    query = request.args.get('query', '')
    mc_version = request.args.get('mc_version')
    loader = request.args.get('loader')
    limit = int(request.args.get('limit', 20))
    offset = int(request.args.get('offset', 0))
    index = request.args.get('index', 'downloads')
    
    result = modrinth_client.search_mods(
        query=query,
        mc_version=mc_version,
        loader=loader,
        limit=limit,
        offset=offset,
        index=index
    )
    return jsonify(result)


@modrinth_bp.route('/mod/<mod_id>', methods=['GET'])
def get_mod(mod_id):
    """Get detailed information about a specific mod."""
    result = modrinth_client.get_mod(mod_id)
    return jsonify(result)


@modrinth_bp.route('/mod/<mod_id>/versions', methods=['GET'])
def get_mod_versions(mod_id):
    """
    Get all versions of a mod.
    
    Query params:
        - loaders: Comma-separated loaders (e.g., "fabric,forge")
        - game_versions: Comma-separated game versions (e.g., "1.20.1,1.20.2")
        - featured: Filter for featured versions (true/false)
    """
    loaders = request.args.getlist('loaders')
    game_versions = request.args.getlist('game_versions')
    featured = request.args.get('featured')
    
    # Convert featured string to boolean if provided
    featured_bool = None
    if featured is not None:
        featured_bool = featured.lower() in ('true', '1', 'yes')
    
    result = modrinth_client.get_mod_versions(
        mod_id=mod_id,
        loaders=loaders if loaders else None,
        game_versions=game_versions if game_versions else None,
        featured=featured_bool
    )
    return jsonify(result)


@modrinth_bp.route('/mod/<mod_id>/download-url', methods=['GET'])
def get_mod_download_url(mod_id):
    """
    Get the direct download URL for a mod's best matching version.
    
    Query params:
        - mc_version: Minecraft version (required, e.g., "1.20.1")
        - loader: Mod loader (default: "fabric")
    """
    mc_version = request.args.get('mc_version')
    loader = request.args.get('loader', 'fabric')
    
    if not mc_version:
        return jsonify({"error": "mc_version parameter is required"}), 400
    
    download_url = modrinth_client.get_mod_download_url(
        mod_id=mod_id,
        mc_version=mc_version,
        loader=loader
    )
    
    if not download_url:
        return jsonify({"error": "No suitable version found"}), 404
    
    return jsonify({"download_url": download_url})


@modrinth_bp.route('/mod/<mod_id>/install', methods=['POST'])
def install_mod(mod_id):
    """
    Download and install a mod to the server's mods folder.
    
    Body params (JSON):
        - mc_version: Minecraft version (required, e.g., "1.20.1")
        - loader: Mod loader (default: "fabric")
        - server_id: Target server ID (optional, defaults to legacy single server)
    """
    data = request.get_json() or {}
    mc_version = data.get('mc_version')
    loader = data.get('loader', 'fabric')
    server_id = data.get('server_id')
    mods_folder_override = data.get('mods_folder')
    
    if not mc_version:
        return jsonify({"error": "mc_version is required"}), 400

    if mods_folder_override:
        return jsonify({"error": "mods_folder override is not allowed"}), 400
    
    mods_folder, error = _resolve_mods_folder(server_id)
    if error:
        payload, status = error
        return jsonify(payload), status
    
    target_path = Path(mods_folder)
    
    # Get download URL
    download_url = modrinth_client.get_mod_download_url(
        mod_id=mod_id,
        mc_version=mc_version,
        loader=loader
    )
    
    if not download_url:
        return jsonify({"error": "No suitable version found"}), 404
    
    # Download the mod
    file_path = modrinth_client.download_mod(download_url, target_path)
    
    if not file_path:
        return jsonify({"error": "Download failed"}), 500
    
    return jsonify({
        "success": True,
        "message": f"Mod installed successfully",
        "file": str(file_path.name),
        "path": str(file_path)
    })


@modrinth_bp.route('/version/<version_id>', methods=['GET'])
def get_version(version_id):
    """Get detailed information about a specific mod version."""
    result = modrinth_client.get_version(version_id)
    return jsonify(result)


@modrinth_bp.route('/categories', methods=['GET'])
def get_categories():
    """Get all available mod categories."""
    result = modrinth_client.get_categories()
    return jsonify(result)


@modrinth_bp.route('/loaders', methods=['GET'])
def get_loaders():
    """Get all available mod loaders."""
    result = modrinth_client.get_loaders()
    return jsonify(result)


@modrinth_bp.route('/game-versions', methods=['GET'])
def get_game_versions():
    """Get all available game versions."""
    result = modrinth_client.get_game_versions()
    return jsonify(result)

File: frontend/src/api/modrinth.js
/**
 * Modrinth API Client
 *
 * One function per panel route. Doc blocks are reserved for behaviour you
 * cannot read off the call itself; a block that restates the signature is one
 * more thing to keep in sync and tells a reader nothing.
 */

import { get, post, del, ApiError } from './client'


export async function searchMods({
  query = '',
  version = '',
  loader = '',
  projectType = 'mod',
  sort = 'relevance',
  limit = 20,
  offset = 0
}, options = {}) {
  return get('/api/modrinth/search', {
    query,
    mc_version: version,
    loader,
    project_type: projectType,
    index: sort,
    limit,
    offset
  }, { signal: options.signal })
}

export async function getModDetails(modId, options = {}) {
  return get(`/api/modrinth/mod/${modId}`, {}, { signal: options.signal })
}

export async function getModVersions(modId, filters = {}, options = {}) {
  return get(`/api/modrinth/mod/${modId}/versions`, filters, { signal: options.signal })
}

/**
 * Install a mod on the server.
 *
 * @param {string} [options.version_id] - Install this exact version instead of
 *   letting the backend resolve the newest compatible one (#56). Recorded as a
 *   pin, so it is distinguishable from an auto-resolved install.
 * @param {string} [options.replaces] - Filename of the jar this supersedes;
 *   removed only after the new one is safely on disk. Must be a bare filename
 *   inside the mods folder. This is what makes a version *change* rather than
 *   a second copy.
 */
export async function installMod(modId, { mc_version, loader, server_id, version_id, replaces }) {
  const body = { mc_version, loader, server_id }
  if (version_id) body.version_id = version_id
  if (replaces) body.replaces = replaces
  return post(`/api/modrinth/mod/${modId}/install`, body)
}

export async function getCategories() {
  return get('/api/modrinth/categories')
}

export async function getLoaders() {
  return get('/api/modrinth/loaders')
}

export async function getGameVersions() {
  return get('/api/modrinth/game-versions')
}

export async function searchModpacks({
  query = '',
  version = '',
  loader = '',
  sort = 'relevance',
  limit = 8,
  offset = 0
}) {
  return get('/api/modrinth/modpacks/search', {
    query,
    mc_version: version,
    loader,
    index: sort,
    limit: Math.min(limit, 50),
    offset
  })
}

export async function getProjectDetails(projectIdOrSlug) {
  return get(`/api/modrinth/project/${encodeURIComponent(projectIdOrSlug)}`)
}

/**
 * Identify every jar in a server's mods folder by content hash.
 *
 * One request for the whole folder. The backend hashes the jars and asks
 * Modrinth's bulk `version_files` endpoint, so this replaces the old
 * per-filename slug guessing that issued ~3.6 requests per jar and tripped
 * Modrinth's rate limit on any hand-populated modpack (#52).
 *
 * Answers `{ resolved }` keyed by filename; jars Modrinth doesn't recognise
 * are simply absent from it.
 */
export async function resolveInstalledMods(serverId, options = {}) {
  return get(
    `/api/modrinth/servers/${encodeURIComponent(serverId)}/resolve-installed`,
    {},
    { signal: options.signal }
  )
}

/**
 * Resolve a project version compatible with the requested game version/loader.
 *
 * @param {string[]} [options.loaders] - Accepted loader facet chain (plugin
 *   servers); when provided it supersedes `loader` server-side.
 */
export async function resolveProjectVersion(projectId, { mc_version, loader, loaders }) {
  const params = { mc_version }
  if (Array.isArray(loaders) && loaders.length) {
    params.loaders = loaders.join(',')
  } else if (loader) {
    params.loader = loader
  }
  return get(`/api/modrinth/project/${encodeURIComponent(projectId)}/resolve-version`, params)
}

export async function getModpackInstallProgress(serverId) {
  return get(`/api/modrinth/modpack/install-progress/${encodeURIComponent(serverId)}`)
}

/** Downloads the .mrpack, installs all server-side mods, and applies overrides. */
export async function installModpack(projectId, {
  mc_version,
  loader,
  server_id,
  clean_install = false,
  create_backup = false,
  allow_missing = false,
  mod_side_overrides = null
}) {
  return post(`/api/modrinth/modpack/${encodeURIComponent(projectId)}/install`, {
    mc_version,
    loader,
    server_id,
    clean_install,
    create_backup,
    allow_missing,
    mod_side_overrides
  })
}

/**
 * Upload a .mrpack exported from the Modrinth app (#53).
 *
 * The archive is staged, not installed: the response describes what the pack
 * declares (`name`, `minecraft_version`, `loader`, ...) so the create form can
 * fill itself in before the target server exists. Install it later with
 * `installUploadedModpack(upload_id, ...)`.
 *
 * Uses XMLHttpRequest rather than fetch for the one thing fetch cannot give
 * us: upload progress events. Same shape as `uploadWorld` in api/backups.js.
 *
 * @param {(pct:number)=>void} [opts.onProgress] 0–100, or -1 when indeterminate
 * @param {(abort:()=>void)=>void} [opts.registerAbort] receives a cancel fn
 */
export function uploadModpackArchive(file, { onProgress, registerAbort } = {}) {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest()
    xhr.open('POST', `/api/modrinth/modpack/upload?filename=${encodeURIComponent(file.name)}`)
    xhr.setRequestHeader('Content-Type', 'application/octet-stream')

    if (typeof registerAbort === 'function') {
      registerAbort(() => xhr.abort())
    }

    xhr.upload.onprogress = (event) => {
      if (typeof onProgress !== 'function') return
      onProgress(event.lengthComputable ? Math.round((event.loaded / event.total) * 100) : -1)
    }

    xhr.onload = () => {
      let data = {}
      try {
        data = JSON.parse(xhr.responseText || '{}')
      } catch (_) {
        data = {}
      }
      if (xhr.status >= 200 && xhr.status < 300) {
        resolve(data)
      } else {
        reject(new ApiError(
          data.error || data.message || `Upload failed with status ${xhr.status}`,
          xhr.status,
          data && Object.keys(data).length ? data : null
        ))
      }
    }

    xhr.onerror = () => reject(new ApiError('Network error during upload', 0, null))
    xhr.onabort = () => reject(new ApiError('Upload cancelled', 0, null))

    xhr.send(file)
  })
}

/**
 * Install a previously uploaded .mrpack onto a server.
 *
 * `force` acknowledges a pack built for a different Minecraft version or
 * loader than the server runs; without it the backend answers 409 with the
 * mismatch spelled out. The upload survives a failed install so the
 * missing-files / uncertain-mod-side retries can reuse it.
 */
export async function installUploadedModpack(uploadId, {
  server_id,
  loader,
  clean_install = false,
  create_backup = false,
  allow_missing = false,
  mod_side_overrides = null,
  force = false
}) {
  return post(`/api/modrinth/modpack/upload/${encodeURIComponent(uploadId)}/install`, {
    server_id,
    loader,
    clean_install,
    create_backup,
    allow_missing,
    mod_side_overrides,
    force
  })
}

/** Drop a staged .mrpack the user decided not to install. */
export async function discardModpackUpload(uploadId) {
  return del(`/api/modrinth/modpack/upload/${encodeURIComponent(uploadId)}`)
}
